### PR TITLE
Fix cross-tab sync for collection-root subscribers

### PR DESCRIPTION
**Pull request by elirangoshen**
### Details

Supersedes https://github.com/Expensify/react-native-onyx/pull/803 (same changes with latest `main` merged in — re-opened from a branch I can maintain). Originally coming from [here](https://github.com/Expensify/react-native-onyx/pull/799#issuecomment-4831583408), authored by @fabioh8010.

Cross-tab (multi-instance) storage sync events were not delivered to collection-root subscribers — e.g. the LHN subscribing to the whole `report_` collection — so actions like pin/unpin, drafting a message, or marking a report read/unread performed in one tab did not sync to other open tabs.

This PR:

- Batches storage sync events and notifies collection-root subscribers (not just individual member-key subscribers) when collection members sync from another instance.
- Handles legacy pre-batching sync events for backward compatibility.
- Splits large sync events into chunks to stay under storage event size limits.

**Note:** this fix depends on the OnyxDerived recompute coalescing that landed in E/App via https://github.com/Expensify/App/pull/96785 (now on staging). Without coalescing, the extra collection-root notifications caused a derived-value recompute storm. Re-verified against latest E/App `main` (which includes #96785) — see Manual Tests.

### Related Issues

https://github.com/Expensify/App/issues/94839

### Linked E/App PR

https://github.com/Expensify/App/pull/97218

### Automated Tests

Added tests in `tests/unit/onyxTest.ts` verifying that both collection-root subscribers (`waitForCollectionCallback: true`) and collection member subscribers are notified when a collection member syncs from another instance. Existing multi-instance sync tests were updated to the new batched event shape.

### Manual Tests

Tested on web against latest E/App `main` (includes the OnyxDerived microtask coalescing from https://github.com/Expensify/App/pull/96785), with Onyx pinned to this PR's head (latest `main` merged in, v3.0.94):

1. Opened two NewDot tabs signed in to the same account.
2. In Tab A: pinned/unpinned a report, drafted a message in a mid-list report, marked a report read/unread, and sent a message — verified each action synced to Tab B (pin icon, LHN re-ordering, draft icon, unread bold, new message) without refreshing.
3. Verified no derived recompute storm: instrumented `OnyxDerived` compute counts in the idle tab — only ~3–6 recomputes per action (previous failure mode without coalescing was thousands of recomputes and a frozen tab).
4. Verified no console errors and no tab freeze in either tab.

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I linked the corresponding Expensify/App PR in the `### Linked E/App PR` section above, and verified this change against it (E/App CI passed and manual testing completed)
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/ae5c35bf-5aa7-4578-9e44-76ce2bc643e9

https://github.com/user-attachments/assets/55585777-a31e-4e1f-86a0-3d21d2e0d37e

https://github.com/user-attachments/assets/bf627541-cc7b-456b-8d03-a59ce553c859

</details>

### CODE REVIEW

**Comment by elirangoshen** [comment]
@codex review

File: lib/storage/InstanceSync/index.web.ts [hunk 1]
```diff
@@ -3,36 +3,101 @@
...
+        if (chunk.length > 0 && chunkLength + keyLength > MAX_SYNC_PAYLOAD_LENGTH) {
+            emitSyncEvent(JSON.stringify(chunk));
+            chunk = [];
+            chunkLength = 2;
+        }
+        chunk.push(onyxKey);
+        chunkLength += keyLength;
     }
+    if (chunk.length > 0) {
+        emitSyncEvent(JSON.stringify(chunk));
```
[the remark this answers is not in the document]
Comment (elirangoshen) [reply lib/storage/InstanceSync/index.web.ts:77]: fixed

**Comment by Beamanator** [comment]
Eyyy i'm traveling a bit today so won't be able to get to this review till tomorrow, @mountiny looks like you're assigned to the issue https://github.com/Expensify/App/issues/94839, do you want to review today?

File: lib/storage/InstanceSync/index.web.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (fabioh8010) [reply lib/storage/InstanceSync/index.web.ts:77]: I dont think we should have applied this – we are adding lots of complexity to "solve" a rare edge case that will happen only once, correct? Doesn't seem worth it to me.
@elirangoshen  Will it crash for the users if we don't apply the fix?

[the remark this answers is not in the document]
Comment (elirangoshen) [reply lib/storage/InstanceSync/index.web.ts:77]: It was working fine before so Ill validate and remove it

**Comment by truph01** [comment]
@mountiny Could you assign me as a reviewer since I was assigned to https://github.com/Expensify/App/issues/94839?

**Comment by mountiny** [comment]
Happy to review, @situchan can you work on the checklist?

**Comment by situchan** [comment]
> @situchan can you work on the checklist?

yes

File: lib/storage/InstanceSync/index.web.ts (hunk 1 above)
Comment (situchan) [inline lib/storage/InstanceSync/index.web.ts:77]: **Chunking can split a collection across two events.**

Chunks are cut purely by payload size, so if one collection's members land in two different chunks, the receiver sees two `storage ` events → two `keysChanged` calls → two collection notifications instead of one.
This partially undoes the O(N²)→O(1) win exactly in the large-batch case chunking exists for.
Grouping by collection before chunking would avoid it.


Not blocker though

File: tests/unit/onyxTest.ts [hunk 2]
```diff
@@ -3298,6 +3298,95 @@ describe('RAM-only keys should not read from storage', () => {
         Onyx.disconnect(connection2);
     });
 
+    it('should notify collection-root and collection member subscribers when a collection member syncs from another instance', async () => {
```
Comment (situchan) [inline tests/unit/onyxTest.ts:3301]: **A couple of test gaps worth filling.**

The two new tests are good, but I'd add:

- A deletion test (member removed in another tab → subscribers get `undefined`). Not only verifying this manually.
- A chunking test around the size limit — that logic is currently untested.
- Optionally, a test for the legacy raw-key path in `parseSyncOnyxStorageEventValue`.

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified that the composer does not automatically focus or open the keyboard on mobile unless explicitly intended. This includes checking that returning the app from the background does not unexpectedly open the keyboard.
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: HybridApp
    - [x] Android: mWeb Chrome
    - [x] iOS: HybridApp
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] For any bug fix or new feature in this PR, I verified that sufficient [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) are included to prevent regressions in this flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: HybridApp</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: HybridApp</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/user-attachments/assets/159def75-ffad-4e58-899e-2f15488c6f52


</details>

File: lib/storage/InstanceSync/index.web.ts (hunk 1 above)
Comment (elirangoshen) [reply lib/storage/InstanceSync/index.web.ts:77]: I think its an edge case , maybe not worth to fix for now.
what do you think @fabioh8010  ?

File: tests/unit/onyxTest.ts (hunk 2 above)
Comment (elirangoshen) [reply tests/unit/onyxTest.ts:3301]: I added few tests now.

File: lib/storage/InstanceSync/index.web.ts (hunk 1 above)
Comment (fabioh8010) [reply lib/storage/InstanceSync/index.web.ts:77]: Yeah I don't think it's harmful to keep current way, the split is supposed to happen only with really big payloads in specific situations.

**Review by fabioh8010 — APPROVED** [review]

**Review by situchan — APPROVED** [review]

File: lib/storage/InstanceSync/index.web.ts [hunk 3]
```diff
@@ -3,36 +3,101 @@
...
+// (and pair it with a try/catch in emitSyncEvent). This way a large key batch (e.g. Onyx.clear() on a
+// heavy account, or a bulk import) is split across several events instead of throwing QuotaExceededError.
+// See https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
+const MAX_SYNC_PAYLOAD_LENGTH = 1_000_000;
+
 /**
- * Raise an event through `localStorage` to let other tabs know a value changed
- * @param {String} onyxKey
+ * Parses the SYNC_ONYX storage event value.
+ * The payload is a JSON array of the changed keys (a batch). I fall backs to treating the raw
```
Comment (mountiny) [inline lib/storage/InstanceSync/index.web.ts]: > I fall backs to treating the raw

?

**Review by mountiny — DISMISSED** [review]
Thanks, make sense!

File: lib/Onyx.ts [hunk 4]
```diff
@@ -21,6 +21,9 @@ import type {
     OnyxInput,
     OnyxMethodMap,
     SetOptions,
+    OnyxCollection,
```
Comment (Beamanator) [inline lib/Onyx.ts]: NAB, could be nice to alphabetize these?

File: lib/Onyx.ts [hunk 5]
```diff
@@ -50,21 +53,54 @@ function init({
...
+                cache.set(key, value);
+
+                if (isCollectionMember && collectionKey) {
+                    let batch = collectionBatches.get(collectionKey);
+                    if (!batch) {
+                        batch = {partial: {}, previous: {}};
+                        collectionBatches.set(collectionKey, batch);
+                    }
+                    batch.partial[key] = value;
+                    // Keep the earliest previous value in case the same member appears twice in one batch.
```
Comment (Beamanator) [inline lib/Onyx.ts:86]: We like adding a blank line between comments and code on the same indentation level 🙏 

```suggestion

                    // Keep the earliest previous value in case the same member appears twice in one batch.
```

**Review by Beamanator — DISMISSED** [review]
Just a few NABs

File: lib/storage/providers/types.ts [hunk 6]
```diff
@@ -99,8 +100,8 @@ type StorageProvider<TStore> = {
     /**
      * @param onStorageKeyChanged Storage synchronization mechanism keeping all opened tabs in sync
```
Comment (Beamanator) [inline lib/storage/providers/types.ts]:
```suggestion
     * @param OnStorageKeysChanged Storage synchronization mechanism keeping all opened tabs in sync
```

File: lib/storage/InstanceSync/index.web.ts (hunk 3 above)
Comment (elirangoshen) [reply lib/storage/InstanceSync/index.web.ts]: fixed the typo

File: lib/Onyx.ts (hunk 4 above)
Comment (elirangoshen) [reply lib/Onyx.ts]: done

File: lib/storage/providers/types.ts (hunk 6 above)
Comment (elirangoshen) [reply lib/storage/providers/types.ts]: fixed

File: lib/Onyx.ts (hunk 5 above)
Comment (elirangoshen) [reply lib/Onyx.ts:86]: fixed

**Review by Beamanator — APPROVED** [review]

**Comment by Beamanator** [comment]
All you @mountiny

**Review by mountiny — APPROVED** [review]
